drivers/BaseCRC.h

```diff
@@ -0,0 +1,238 @@
+     */
+    virtual crc_polynomial_t get_polynomial(void) const
+    {
+        crc_polynomial_type_t polynomial = this->get_polynomial_type();
+
+        if (CRC_7BIT_SD == polynomial) {
+            return POLY_7BIT_SD;
+        } else if (CRC_8BIT_CCITT == polynomial) {
+            return POLY_8BIT_CCITT;
+        } else if ((CRC_16BIT_SD == polynomial) || (CRC_16BIT_CCITT == polynomial)) {
+            return POLY_16BIT_CCITT;
+        } else if (CRC_16BIT_IBM == polynomial) {
+            return POLY_16BIT_IBM;
+        } else if (CRC_32BIT == polynomial) {
+            return POLY_32BIT_ANSI;
+        }
+        return POLY_INVALID;
+    }
+
+    /** Get the current CRC width
+     *
+     * @return  CRC width, 0 when polynomial type is incorrect or not supported.
+     */
+    virtual crc_width_t get_width(void) const
+    {
+        crc_polynomial_type_t polynomial = this->get_polynomial_type();
+
+        if (CRC_7BIT_SD == polynomial) {
+            return CRC_WIDTH_7;
+        } else if (CRC_8BIT_CCITT == polynomial) {
+            return CRC_WIDTH_8;
+        } else if ((CRC_16BIT_CCITT == polynomial) ||
+                (CRC_16BIT_SD == polynomial) ||
+                (CRC_16BIT_IBM == polynomial)) {
+            return CRC_WIDTH_16;
+        } else if (CRC_32BIT == polynomial) {
+            return CRC_WIDTH_32;
+        }
+        return CRC_WIDTH_INVALID;
+    }
+
+    /** Compute CRC for the data input
+     *
+     *  @param   buffer  Data bytes
+     *  @param   size  Size of data
+     *  @param  crc  CRC is the output value
+     *  @return  0 on success, negative error code on failure
+     */
+    virtual int32_t compute(void *buffer, crc_data_size_t size, uint32_t *crc) = 0;
+
+    /** Compute partial start, indicate start of partial computation
+     *
+     *  This API should be called before performing any partial computation
+     *  with compute_partial API.
+     *
+     *  @param  crc  Initial CRC value is the output parameter, value is set by the API
+     *  @return  0  on success or a negative when not supported
+     *  @note: CRC is an out parameter and must be reused with compute_partial
+     *         and compute_partial_stop without any modifications in application.
+     */
+    virtual int32_t compute_partial_start(uint32_t *crc)
+    {
+        return -1;
+    }
+
+    /** Compute partial CRC for the data input.
+     *
+     *  CRC data if not available fully, CRC can be computed in parts with available data.
+     *  Previous CRC output should be passed as argument to the current compute_partial call.
+     *  @pre: Call \ref compute_partial_start to start the partial CRC calculation.
+     *  @post: Call \ref compute_partial_stop to get the final CRC value.
+     *
+     *  @param  buffer  Data bytes
+     *  @param  size  Size of data
+     *  @param  crc  CRC value is intermediate CRC value filled by API.
+     *  @return  0  on success or a negative error code on failure
```

Comment:
Shall we define negative error codes, or -1 does it all?

Reply:
-1 is the only error code in Software CRC implementation. However, I can use "Not supported" as additional error code, if we have common error codes, but nothing special at present.